File: src/model/item.cpp
//
// Created by samuel on 18/12/22.
//


#include <btBulletDynamicsCommon.h>
#include <glm/gtc/quaternion.hpp>
#include <glm/gtc/type_ptr.hpp>
#include <iostream>
#include <utility>

#include "./converter.h"
#include "item.h"

Item::Item(
    std::string name, const std::shared_ptr<Shape> &shape, glm::mat4 model_matrix, glm::vec3 scale,
    float mass)
    : name(std::move(name)), shape(shape), scale(scale), first_model_matrix(model_matrix) {
    auto *convex_hull_shape = new btConvexHullShape();

    for (auto [x, y, z]: shape->get_vertices()) convex_hull_shape->addPoint(btVector3(x, y, z));

    collision_shape = convex_hull_shape;

    collision_shape->setLocalScaling(glm_to_bullet(scale));

    btVector3 local_inertia(0, 0, 0);
    if (mass != 0.f) collision_shape->calculateLocalInertia(mass, local_inertia);

    btTransform original_tr;
    original_tr.setFromOpenGLMatrix(glm::value_ptr(model_matrix));

    auto *motion_state = new btDefaultMotionState(original_tr);

    btRigidBody::btRigidBodyConstructionInfo body_info(
        mass, motion_state, collision_shape, local_inertia);

    body = new btRigidBody(body_info);
}

Item::Item(
    std::string name, const std::shared_ptr<Shape> &shape, glm::vec3 position, glm::quat rotation,
    glm::vec3 scale, float mass)
    : Item(
          std::move(name), shape,
          glm::translate(glm::mat4(1.f), position) * glm::mat4_cast(rotation), scale, mass) {}

Item::Item(
    std::string name, const std::shared_ptr<Shape> &shape, glm::vec3 position, glm::vec3 scale,
    float mass)
    : Item(std::move(name), shape, position, glm::quat_cast(glm::mat4(1.f)), scale, mass) {}

std::shared_ptr<Shape> Item::get_shape() { return shape; }

std::string Item::get_name() { return name; }

btRigidBody *Item::get_body() { return body; }

glm::mat4 Item::model_matrix() {
    return model_matrix_without_scale() * glm::scale(glm::mat4(1.f), scale);
}

glm::mat4 Item::model_matrix_without_scale() {
    btTransform tr;
    body->getMotionState()->getWorldTransform(tr);
    return bullet_to_glm(tr);
}

void Item::reset(glm::mat4 main_model_matrix) {
    btTransform original_tr = glm_to_bullet(main_model_matrix * first_model_matrix);

    body->setWorldTransform(original_tr);
    body->getMotionState()->setWorldTransform(original_tr);

    body->setLinearVelocity(btVector3(0.f, 0.f, 0.f));
    body->setAngularVelocity(btVector3(0.f, 0.f, 0.f));
    body->clearForces();
}

File: evo_motion/model/item.py
import pybullet as pb
from glm import mat4, mat4_cast, quat
from glm import scale as glm_scale
from glm import translate, vec3

from .shapes import Shape


class Item:
    def __init__(
        self,
        name: str,
        shape: Shape,
        position: vec3,
        scale: vec3,
        mass: float,
    ) -> None:

        self.__name = name
        self.__shape = shape

        self.__shape_id: int = pb.createCollisionShape(
            pb.GEOM_MESH, vertices=shape.get_vertices(), meshScale=scale
        )

        self.__body_id: int = pb.createMultiBody(
            baseCollisionShapeIndex=self.__shape_id,
            basePosition=list(position),
            baseMass=mass,
        )

        self.__scale = scale

    @property
    def name(self) -> str:
        return self.__name

    @property
    def body(self) -> int:
        return self.__body_id

    @property
    def collision_shape(self) -> int:
        return self.__shape_id

    def shape(self) -> Shape:
        return self.__shape

    def model_matrix(self) -> mat4:
        position, quaternion = pb.getBasePositionAndOrientation(self.__body_id)

        scale_mat: mat4 = glm_scale(mat4(1), self.__scale)
        rot_matrix: mat4 = mat4_cast(quat(quaternion))
        trans_matrix: mat4 = translate(
            mat4(1), vec3(position[0], position[1], position[2])
        )

        model_matrix: mat4 = trans_matrix * rot_matrix * scale_mat

        return model_matrix
